index.js

```diff
@@ -91,6 +91,11 @@ async function rateLimitPlugin (fastify, settings) {
   pluginComponent.run = run
   fastify.decorateRequest(run, false)
 
+  if (!fastify.hasDecorator('rateLimit')) {
+    // The rate limit plugin can be registered multiple times but decorate throws if called multiple times for the same field
+    fastify.decorate('rateLimit', rateLimitRequestHandler(globalParams, pluginComponent))
+  }
```

Comment:
There are 2 things I don't like with this implementation but it seems to be what was suggested in issue #127:
1. The rateLimit uses the global params without the ability to customize that at the `preHandler` level of the `setNotFoundHandler`
2. The plugin can supposedly be registered more than once but this implementation will only retain the first registration to the decorator.

Reply:
Yes good point. Like you suggested, I made `rateLimit` a function so that it makes the signature more forward compatible.

Reply:
> Yes good point. Like you suggested, I made rateLimit a function so that it makes the signature more forward compatible.

The current code does not match what I recommended.

Reply:
use it as function should allow user to pass in `options` and which they can customize the `rateLimitRequestHandler`. I think it can be extracted to be a big `createRateLimit` function and which you can use it in both `plugin` and `decorators`.

Reply:
@mcollina @climba03003 ok so if I understand what you are saying is that you would like the configuration `fastify.rateLimit(opts)` to be part of this PR as opposed to having an empty placeholder `fastify.rateLimit(void)` for now ?

Reply:
Yes, it is.